File: scripts/swe-bench/official-harness.ts
import { join } from "node:path";
import type { SecretRedactor } from "./redactor.ts";
import type { SweBenchTask } from "./types.ts";

interface HarnessResponse {
  readonly evalScript?: string;
  readonly [key: string]: unknown;
}

/** 通过固定 Python venv 调用官方 SWE-bench 代码，不在 TypeScript 重写判分。 */
export class OfficialHarness {
  readonly #python: string;
  readonly #helper: string;
  readonly #redactor: SecretRedactor;

  /** 绑定仓库 cache 中的固定 venv 与窄 Python adapter。 */
  constructor(repositoryRoot: string, redactor: SecretRedactor) {
    this.#python = join(repositoryRoot, ".cache", "swe-bench", "venv", "bin", "python");
    this.#helper = join(repositoryRoot, "scripts", "swe-bench", "official_harness.py");
    this.#redactor = redactor;
  }

  /** 使用官方 TestSpec 生成含测试退出码标记的 eval script。 */
  async prepare(task: SweBenchTask, image: string): Promise<string> {
    const response = await this.#invoke({ action: "prepare", task: { ...task, image } });
    if (typeof response.evalScript !== "string")
      throw new Error("official harness returned no eval script");
    return response.evalScript;
  }

  /** 使用官方 log parser 和 grading 生成逐测试报告。 */
  async grade(
    task: SweBenchTask,
    image: string,
    modelPatch: string,
    evaluationLog: string,
  ): Promise<unknown> {
    return this.#invoke({
      action: "grade",
      task: { ...task, image },
      modelPatch,
      evaluationLog,
    });
  }

  /** 通过 stdin 传请求，避免题面或 patch 进入宿主命令行。 */
  async #invoke(payload: unknown): Promise<HarnessResponse> {
    const child = Bun.spawn([this.#python, this.#helper], {
      stdin: "pipe",
      stdout: "pipe",
      stderr: "pipe",
      env: { ...Bun.env, PYTHONNOUSERSITE: "1" },
    });
    child.stdin.write(JSON.stringify(payload));
    child.stdin.end();
    const [stdout, stderr, code] = await Promise.all([
      new Response(child.stdout).text(),
      new Response(child.stderr).text(),
      child.exited,
    ]);
    if (code !== 0) {
      throw new Error(`official SWE-bench harness failed: ${this.#redactor.redact(stderr)}`);
    }
    return JSON.parse(stdout) as HarnessResponse;
  }
}

File: scripts/swe-bench/official_harness.py
#!/usr/bin/env python3
"""固定调用官方 SWE-bench TestSpec、eval-script 与 grading 的窄适配器。"""

from __future__ import annotations

import json
import sys
from pathlib import Path

from swebench.harness.grading import get_eval_report
from swebench.harness.utils import make_test_spec


def load_payload() -> dict:
    """从 stdin 读取 TypeScript runner 提供的单个 JSON 请求。"""
    return json.load(sys.stdin)


def instance_from_task(task: dict) -> dict:
    """把仓库 task schema 映射为官方 make_test_spec 所需字段。"""
    return {
        "instance_id": task["id"],
        "image": task["image"],
        "repo": task["repo"],
        "version": task["version"],
        "FAIL_TO_PASS": task["failToPass"],
        "PASS_TO_PASS": task["passToPass"],
        "log_parser": task["logParser"],
        "eval_type": task["evalType"],
        "eval_script": task["evalScript"],
    }


def main() -> int:
    """执行 prepare 或 grade，判分真值完全交给固定版本官方包。"""
    payload = load_payload()
    test_spec = make_test_spec(instance_from_task(payload["task"]))
    action = payload["action"]
    if action == "prepare":
        json.dump({"evalScript": test_spec.eval_script}, sys.stdout)
        return 0
    if action == "grade":
        prediction = {
            "instance_id": test_spec.instance_id,
            "model_name_or_path": "minicode",
            "model_patch": payload["modelPatch"],
        }
        report = get_eval_report(
            test_spec,
            prediction,
            str(Path(payload["evaluationLog"])),
            include_tests_status=True,
        )
        json.dump(report, sys.stdout)
        return 0
    raise ValueError(f"unsupported action: {action}")


if __name__ == "__main__":
    raise SystemExit(main())
